File: appmonet-sdk/appmonet-sdk-core/src/main/java/com/monet/bidder/ApplicationCallbacks.java
package com.monet.bidder;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import com.monet.bidder.auction.AuctionManager;

/**
 * Listen to application lifecycle events
 */
class ApplicationCallbacks implements Application.ActivityLifecycleCallbacks {
  private final AuctionManager auctionManager;
  private int mActiveActivities = 0;

  ApplicationCallbacks(AuctionManager auctionManager) {
    this.auctionManager = auctionManager;
  }

  private void trackAppState(String appStateChange, String identifier) {
    auctionManager.trackAppState(appStateChange, identifier);
  }

  @Override
  public void onActivityCreated(Activity activity, Bundle bundle) {
    trackAppState(Constants.JSAppStates.ACTIVITY_CREATED, activity.getLocalClassName());
  }

  @Override
  public void onActivityStarted(Activity activity) {
    trackAppState(Constants.JSAppStates.ACTIVITY_STARTED, activity.getLocalClassName());
    if (mActiveActivities == 0) {
      trackAppState(Constants.JSAppStates.FOREGROUND, "app");
    }
    mActiveActivities += 1;
  }

  @Override
  public void onActivityResumed(Activity activity) {
    trackAppState(Constants.JSAppStates.ACTIVITY_RESUMED, activity.getLocalClassName());
  }

  @Override
  public void onActivityPaused(Activity activity) {
    trackAppState(Constants.JSAppStates.ACTIVITY_PAUSED, activity.getLocalClassName());
  }

  @Override
  public void onActivityStopped(Activity activity) {
    trackAppState(Constants.JSAppStates.ACTIVITY_STOPPED, activity.getLocalClassName());
    mActiveActivities -= 1;
    if (mActiveActivities == 0) {
      trackAppState(Constants.JSAppStates.BACKGROUND, "app");
    }
  }

  @Override
  public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle bundle) {
    //not implemented
  }

  @Override
  public void onActivityDestroyed(Activity activity) {
    trackAppState(Constants.JSAppStates.ACTIVITY_DESTROYED, activity.getLocalClassName());
  }
}

File: appmonet-sdk/appmonet-sdk-core/src/main/java/com/monet/bidder/ApplicationCallbacks.kt
package com.monet.bidder

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle
import com.monet.bidder.Constants.JSAppStates
import com.monet.bidder.auction.AuctionManager

/**
 * Listen to application lifecycle events
 */
internal class ApplicationCallbacks(private val auctionManager: AuctionManager) :
    ActivityLifecycleCallbacks {
  private var mActiveActivities = 0
  private fun trackAppState(
    appStateChange: String,
    identifier: String
  ) {
    auctionManager.trackAppState(appStateChange, identifier)
  }

  override fun onActivityCreated(
    activity: Activity,
    bundle: Bundle
  ) {
    trackAppState(JSAppStates.ACTIVITY_CREATED, activity.localClassName)
  }

  override fun onActivityStarted(activity: Activity) {
    trackAppState(JSAppStates.ACTIVITY_STARTED, activity.localClassName)
    if (mActiveActivities == 0) {
      trackAppState(JSAppStates.FOREGROUND, "app")
    }
    mActiveActivities += 1
  }

  override fun onActivityResumed(activity: Activity) {
    trackAppState(JSAppStates.ACTIVITY_RESUMED, activity.localClassName)
  }

  override fun onActivityPaused(activity: Activity) {
    trackAppState(JSAppStates.ACTIVITY_PAUSED, activity.localClassName)
  }

  override fun onActivityStopped(activity: Activity) {
    trackAppState(JSAppStates.ACTIVITY_STOPPED, activity.localClassName)
    mActiveActivities -= 1
    if (mActiveActivities == 0) {
      trackAppState(JSAppStates.BACKGROUND, "app")
    }
  }

  override fun onActivitySaveInstanceState(
    activity: Activity,
    bundle: Bundle
  ) {
    //not implemented
  }

  override fun onActivityDestroyed(activity: Activity) {
    trackAppState(JSAppStates.ACTIVITY_DESTROYED, activity.localClassName)
  }
}